File: app/src/main/java/com/byd/dashcast/ui/main/UsageTracker.java
package com.byd.dashcast.ui.main;

import android.content.Context;
import android.content.SharedPreferences;

import com.byd.dashcast.R;
import com.byd.dashcast.data.prefs.ClusterPrefs;

/**
 * Tracks how long each cluster app is displayed and provides a summary dialog.
 * State ({@code mStartTime}) and persistence ({@code usage_ms_<pkg>} in ClusterPrefs)
 * are fully owned here; MainActivity just calls {@link #trackStart()}, {@link #trackStop},
 * and {@link #showStatsDialog()}.
 */
public final class UsageTracker {

    public interface Host {
        Context getContext();
    }

    private final Host mHost;
    private long mStartTime = 0;

    public UsageTracker(Host host) {
        mHost = host;
    }

    public void trackStart() {
        mStartTime = System.currentTimeMillis();
    }

    public void trackStop(String pkgName) {
        if (mStartTime <= 0 || pkgName == null) return;
        long elapsed = System.currentTimeMillis() - mStartTime;
        mStartTime = 0;
        if (elapsed < 1000) return;
        SharedPreferences prefs = mHost.getContext()
                .getSharedPreferences(ClusterPrefs.PREFS_NAME, Context.MODE_PRIVATE);
        long prev = prefs.getLong("usage_ms_" + pkgName, 0);
        prefs.edit().putLong("usage_ms_" + pkgName, prev + elapsed).apply();
    }

    public void showStatsDialog() {
        Context ctx = mHost.getContext();
        SharedPreferences prefs = ctx.getSharedPreferences(ClusterPrefs.PREFS_NAME, Context.MODE_PRIVATE);
        java.util.Map<String, ?> all = prefs.getAll();
        java.util.List<String[]> stats = new java.util.ArrayList<>();
        for (java.util.Map.Entry<String, ?> entry : all.entrySet()) {
            if (entry.getKey().startsWith("usage_ms_") && entry.getValue() instanceof Long) {
                String pkg = entry.getKey().substring("usage_ms_".length());
                long ms = (Long) entry.getValue();
                String name = pkg;
                try {
                    android.content.pm.ApplicationInfo ai =
                            ctx.getPackageManager().getApplicationInfo(pkg, 0);
                    CharSequence label = ctx.getPackageManager().getApplicationLabel(ai);
                    if (label != null) name = label.toString();
                } catch (Exception ignored) {}
                stats.add(new String[]{name, formatDuration(ms)});
            }
        }
        if (stats.isEmpty()) {
            android.widget.Toast.makeText(ctx, ctx.getString(R.string.usage_empty),
                    android.widget.Toast.LENGTH_SHORT).show();
            return;
        }
        java.util.Collections.sort(stats, (a, b) -> a[0].compareToIgnoreCase(b[0]));
        StringBuilder sb = new StringBuilder();
        for (String[] s : stats) sb.append(s[0]).append(" — ").append(s[1]).append("\n");

        new android.app.AlertDialog.Builder(ctx)
                .setTitle(ctx.getString(R.string.usage_title))
                .setMessage(sb.toString().trim())
                .setPositiveButton(android.R.string.ok, null)
                .setNeutralButton(ctx.getString(R.string.usage_reset), (d, w) -> {
                    SharedPreferences.Editor editor = prefs.edit();
                    for (String key : prefs.getAll().keySet()) {
                        if (key.startsWith("usage_ms_")) editor.remove(key);
                    }
                    editor.apply();
                    android.widget.Toast.makeText(ctx,
                            ctx.getString(R.string.toast_usage_stats_reset),
                            android.widget.Toast.LENGTH_SHORT).show();
                })
                .show();
    }

    private static String formatDuration(long ms) {
        long seconds = ms / 1000;
        long minutes = seconds / 60;
        long hours   = minutes / 60;
        if (hours   > 0) return hours   + "h " + (minutes % 60) + "m";
        if (minutes > 0) return minutes + "m " + (seconds % 60) + "s";
        return seconds + "s";
    }
}

File: app/src/main/java/com/byd/dashcast/ui/main/UsageTracker.kt
package com.byd.dashcast.ui.main

import android.app.AlertDialog
import android.content.Context
import android.widget.Toast
import androidx.core.content.edit
import com.byd.dashcast.R
import com.byd.dashcast.data.prefs.ClusterPrefs

/**
 * Tracks how long each cluster app is displayed and provides a summary dialog.
 * State ({@code mStartTime}) and persistence ({@code usage_ms_<pkg>} in ClusterPrefs)
 * are fully owned here; MainActivity just calls {@link #trackStart()}, {@link #trackStop},
 * and {@link #showStatsDialog()}.
 */
class UsageTracker(private val mHost: Host) {

    interface Host {
        fun getContext(): Context
    }

    private var mStartTime: Long = 0

    fun trackStart() {
        mStartTime = System.currentTimeMillis()
    }

    fun trackStop(pkgName: String?) {
        if (mStartTime <= 0 || pkgName == null) return
        val elapsed = System.currentTimeMillis() - mStartTime
        mStartTime = 0
        if (elapsed < 1000) return
        val prefs = mHost.getContext()
            .getSharedPreferences(ClusterPrefs.PREFS_NAME, Context.MODE_PRIVATE)
        val prev = prefs.getLong("usage_ms_$pkgName", 0)
        prefs.edit { putLong("usage_ms_$pkgName", prev + elapsed) }
    }

    fun showStatsDialog() {
        val ctx = mHost.getContext()
        val prefs = ctx.getSharedPreferences(ClusterPrefs.PREFS_NAME, Context.MODE_PRIVATE)
        val all = prefs.all
        val stats = ArrayList<Array<String>>()
        for (entry in all.entries) {
            val value = entry.value
            if (entry.key.startsWith("usage_ms_") && value is Long) {
                val pkg = entry.key.substring("usage_ms_".length)
                val ms = value
                var name = pkg
                try {
                    val ai = ctx.packageManager.getApplicationInfo(pkg, 0)
                    val label: CharSequence? = ctx.packageManager.getApplicationLabel(ai)
                    if (label != null) name = label.toString()
                } catch (ignored: Exception) {
                }
                stats.add(arrayOf(name, formatDuration(ms)))
            }
        }
        if (stats.isEmpty()) {
            Toast.makeText(
                ctx, ctx.getString(R.string.usage_empty),
                Toast.LENGTH_SHORT
            ).show()
            return
        }
        stats.sortWith { a, b -> a[0].compareTo(b[0], ignoreCase = true) }
        val sb = StringBuilder()
        for (s in stats) sb.append(s[0]).append(" — ").append(s[1]).append("\n")

        AlertDialog.Builder(ctx)
            .setTitle(ctx.getString(R.string.usage_title))
            .setMessage(sb.toString().trim())
            .setPositiveButton(android.R.string.ok, null)
            .setNeutralButton(ctx.getString(R.string.usage_reset)) { _, _ ->
                prefs.edit {
                    for (key in prefs.all.keys) {
                        if (key.startsWith("usage_ms_")) remove(key)
                    }
                }
                Toast.makeText(
                    ctx,
                    ctx.getString(R.string.toast_usage_stats_reset),
                    Toast.LENGTH_SHORT
                ).show()
            }
            .show()
    }

    companion object {
        private fun formatDuration(ms: Long): String {
            val seconds = ms / 1000
            val minutes = seconds / 60
            val hours = minutes / 60
            if (hours > 0) return "${hours}h ${minutes % 60}m"
            if (minutes > 0) return "${minutes}m ${seconds % 60}s"
            return "${seconds}s"
        }
    }
}
